### PR TITLE
8769-user-keybinding: propagating Theia overriden keybindings into VS Code editor

### LINKED ISSUE #8769: User keybindings are not propagated into vscode editor
### Bug Description:
Setting User keyboard preferences in Theia is not propagated to vscode editor - there the default keybindings are used.

### Steps to Reproduce:

After PR [8589](https://github.com/eclipse-theia/theia/pull/8589) is merged:
a. Update Keyboard shortcuts for "acceptRenameInput"
b. click F2 to rename a function - rename input field is displayed with updated keyboard settings

** _getResolver is responsible for merging default keybindings with overriden ones (by user keybindings)
image

### Additional Information

- Operating System: tested on windows but relevant to all.
- Theia Version: 1.7.0

### PR DESCRIPTION
**Pull request by danarad05**
Signed-off-by: Dan Arad <dan.arad@sap.com>

#### What it does
Fixes #8769  
VS Code/monaco editor uses default vscode keybindings if no dynamicKeybindings are set. Theia keybindings overriden by user were not propagated into VS Code because they weren't added to dynamicKeybindings array using addDynamicKeybinding method.

On VS Code' side,  **_getResolver** is responsible for merging default keybindings with overriden ones ( = dynamicKeybindings)
image

#### How to test
After PR [8589](https://github.com/eclipse-theia/theia/pull/8589) is merged:
a. Update Keyboard shortcuts for "acceptRenameInput" & "acceptRenameInputWithPreview":
image
b. click F2 to rename a function - rename input field is displayed with updated keyboard settings
image

#### Review checklist

- [X] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/monaco/src/browser/monaco-editor-provider.ts
Comment (danarad05) [inline packages/monaco/src/browser/monaco-editor-provider.ts:199]: I want to put attention to this line. 
It was added in order to be able to run addDynamicKeybinding where vscode has a check concerning it [there](https://github.com/microsoft/vscode/blob/c47d83b293181d9be64f27ff093689e8e7aed054/src/vs/editor/standalone/browser/simpleServices.ts#L344):
Not sure that is best solution but I couldn't find an alternative.
image

**Comment by danarad05** [comment]
@amiramw @vince-fugnitto

[8589](https://github.com/eclipse-theia/theia/pull/8589)  was merged, could we advance this PR?

**Review by vince-fugnitto — COMMENTED** [review]
@danarad05 thank you for the changes, do you mind adding information in both the commit and the pull-request as to what the actual issue is, and how this pull-request resolves it (at the moment there is no information). For example, explain why the keybindings are not propagated to the monaco-editor, and how this pull-request resolves it. 

One thing I do not understand by reading the issue and the pull-request is how some monaco commands (such as `format document`) can be overidden today on master but https://github.com/eclipse-theia/theia/issues/8769 seems to imply it does not?

**Comment by danarad05** [comment]
> @danarad05 thank you for the changes, do you mind adding information in both the commit and the pull-request as to what the actual issue is, and how this pull-request resolves it (at the moment there is no information). For example, explain why the keybindings are not propagated to the monaco-editor, and how this pull-request resolves it.

Please check if updated description is sufficient 

> One thing I do not understand by reading the issue and the pull-request is how some monaco commands (such as `format document`) can be overidden today on master but #8769 seems to imply it does not?

This PR addresses keybindings and not commands.

**Comment by danarad05** [comment]
@vince-fugnitto Where are we standing with this PR?

**Review by vince-fugnitto — COMMENTED** [review]
The updated keybindings only take effect for me when restarting the application, else the <kbd>F2</kbd> (rename) displays the previous value. Can you confirm if you also experience this issue?

- press <kbd>F2</kbd> on method - prompt works and displays proper keybindings
- update the keybindings for both `acceptRenameInput` and `acceptRenameInputWithPreview`
- press <kbd>F2</kbd> on method - prompt displays previous keybindings
- reload application
- press <kbd>F2</kbd> on method - prompt now displays correct keybinding

**Comment by danarad05** [comment]
@vince-fugnitto 

> The updated keybindings only take effect for me when restarting the application, else the F2 (rename) displays the previous value. Can you confirm if you also experience this issue?

1. Sorry for delay in response
2. Rebased
3. Dynamic key bindings are updated when editor it opened. Therefore, after you update shortcut keys you need to reopen editor or open a new editor. Tested and in this way prompt is updated.
4. In continuation of my remark [here](https://github.com/eclipse-theia/theia/pull/9154#issuecomment-844952500) - I would like to complete this PR review ASAP.

Thanks

**Comment by vince-fugnitto** [comment]
> 3. Dynamic key bindings are updated when editor it opened. Therefore, after you update shortcut keys you need to reopen editor or open a new editor. Tested and in this way prompt is updated.

Do you think we should update the dynamic keybindings if we detect they have been changed (similarly to vscode)?
Not handling this use-case means that the `rename input` (as an example) does not display the proper keybinding while new editors do. For an end-user they would be confused as to why one editor displays such a keybinding but not the other(s).

Besides that comment the changes do seem to work as expected.

**Comment by danarad05** [comment]
@vince-fugnitto 

> Do you think we should update the dynamic keybindings if we detect they have been changed (similarly to vscode)?

You are right. added listener to changed keybindings and not no need to reopen editor.

**Review by vince-fugnitto — APPROVED** [review]
I do not have any additional comments regarding the code or behavior :+1: 
I confirmed that:
- the new keybinding is applied for the `preview rename`
- the new keybinding is applied automatically when set, no need to re-open the editor
- the new keybinding is applied successfully when the application is reloaded or restarted

**Comment by danarad05** [comment]
> I do not have any additional comments regarding the code or behavior 👍
> I confirmed that:
> 
> * the new keybinding is applied for the `preview rename`
> * the new keybinding is applied automatically when set, no need to re-open the editor
> * the new keybinding is applied successfully when the application is reloaded or restarted

Thanks @vince-fugnitto. Now we need to find another reviewer?

**Comment by vince-fugnitto** [comment]
> Thanks @vince-fugnitto. Now we need to find another reviewer?

@danarad05 I'm not sure anyone else will be interested in reviewing, is the pull-request urgent or do we wait for after tomorrow's release to merge to be cautious?

**Comment by danarad05** [comment]
> > Thanks @vince-fugnitto. Now we need to find another reviewer?
> 
> @danarad05 I'm not sure anyone else will be interested in reviewing, is the pull-request urgent or do we wait for after tomorrow's release to merge to be cautious?

It's not urgent urgent. Are we not merging any more PRs for this version anymore?  There are other PRs I have that are approved ...

**Comment by vince-fugnitto** [comment]
> It's not urgent urgent. Are we not merging any more PRs for this version anymore? There are other PRs I have that are approved ...

@danarad05 others can be merged, since this one affected monaco I wanted to be a bit more cautious and merge early after the release giving us the chance to identify any issues throughout the month (although I do not expect any).

**Comment by EstherPerelman** [comment]
@msujew @vince-fugnitto and any available reviewer... I have rebased this PR and it's ready for recheck...

**Review by msujew — COMMENTED** [review]
The behavior looks good for me, except for one minor issue: When reloading the window while focusing the `Keyboard shortcuts` view, the view displays the original keybindings. The modified keybindings work as expected, but aren't shown in the view. Closing and reopening the view fixes the issue, but I would still like to see this addressed (as it is quite confusing/misleading for the user).

In my test, I modified the `Cut` command to use `ctrlcmd+m` and reloaded the page without leaving the view. The view still says `ctrlcmd+c`, but using `ctrlcmd+m` still works in the editor.

**Comment by JonasHelming** [comment]
@EstherPerelman : Could you have a look at the issue described above?

**Comment by EstherPerelman** [comment]
> @EstherPerelman : Could you have a look at the issue described above?

@JonasHelming we are in SAP busy with other work now and no more need this PR (We keep a stable version of theia),
Maybe you would like to merge this PR and then to open a separate issue for the issue @msujew found, or take this PR and solve it...
Good Luck

**Comment by JonasHelming** [comment]
@vince-fugnitto @msujew : What is your opinion on the proposal above?

**Comment by msujew** [comment]
@JonasHelming I would be alright with merging it first, before addressing the issue. But we should proceed with the monaco uplift first.

**Pull request closed without merging**
